platform/engineering/design-docs/2020-08-17-mobile-api-design-doc.md

```diff
@@ -0,0 +1,173 @@
+mobile app for the VA. It was immediately apparent that the authenticated Veteran experience would provide the most value, 
+similar to the VA.gov experience. The goal of the mobile app is not to replicate every bit of VA.gov, but to provide a 
+new mode of communication to allow Veterans to interact more easily with the department. Since VA.gov is powered by 
+vets-api and VSP, it seemed like a natural fit for the mobile app to be powered by APIs in the same space.
+
+### High Level Design
+Since vets-api already contains integrations to backend services that the mobile app will need 
+(such as MVI, Claims, Appointments), the mobile API will be built inside of the vets-api codebase, and exposed in the same 
+ way that VA.gov API endpoints are exposed.
+
+The mobile API will provide functionality that is currently exclusive to VA.gov as well as some already-public functionality 
+from Lighthouse, so existing integrations will be reused. A Rails engine will house the mobile API surface area, with necessary 
+functionality connected as closely as possible to the backend integrations. In some cases, this may be a simple passthrough, 
+while in other cases the mobile API will provide its own presentation layer on top of the 
+integrations. This will allow the mobile API to grow independently of the rest of vets-api, while providing a consistent API 
+that will be useful for future mobile applications to consume.
+
+All endpoints will be versioned, so that long-lasting mobile releases will continue to function into the future. When 
+modifications are required, changes will be made in a way to not break existing clients. The team is considering 
+building a machine-readable discovery API endpoint to decouple the mobile app from hard-coded urls.
+
+All mobile authorization will be confined to this Rails engine. The mobile app will require some new endpoints as well, and 
+these will be built in the Rails engine for isolation. [A companion auth design doc has more details](https://github.com/department-of-veterans-affairs/va.gov-team/pull/12482).
+
+## Specifics
+
+### Detailed Design
+The mobile endpoints need custom authentication, merged responses, and reduced payloads. With versioning this is possible 
+in the core api albeit with a fractured experience. Mobile endpoints could be on v0, v1, or even v2. MVP mobile endpoints 
+will be in flux, and parallel va.gov changes could cause versions to climb.
+
+To limit the disruption to the API it's best to perform the optimizations in a module with a `/mobile` base path. 
+This allows us to:
+
+- Add mobile authentication without changes to the core API codebase.
+- Stay out of the way as we tweak endpoints in a mobile sandbox.
+- Test out optimizations that can be later ported back to the core API.
+
+Expanding on the last point it's possible that once the mobile optimized versions of the endpoints have stabilized that we permanently
```

Comment:
can you say more about API/endpoint versioning?

one of my biggest concerns is about mobile clients adding constraints to vets-api over time. mobile clients are likely to permanently lag behind in updates in ways that client JS does not. some questions that come to mind, though I think this is a bigger area to discuss than just these immediate questions:

* how will we know which endpoints/fields/etc. are depended on by mobile clients in the wild, and therefore can't (ever?) be changed?
* because of this concern, should mobile endpoints remain separated from web endpoints? why or why not? (I don't have a strong opinion here, but I think we should discuss pros/cons)
* how might a discovery API facilitate flexibility here?
* will we have a deprecation policy for mobile endpoints, slash, how out-of-date can a user's VA.gov mobile app be before we stop supporting it?
* follow-up: assuming we do eventually stop supporting old versions of mobile apps, will that include functionality in the mobile app to alert the user that they need to upgrade? (i.e. can we control that from an API endpoint of ours?)

Reply:
@gunsch answers from @jjulian and myself for the above questions:

> How will we know which endpoints/fields/etc. are depended on by mobile clients in the wild, and therefore can’t (ever?) be changed? - because of this concern, should mobile endpoints remain separated from web endpoints? why or why not? (I don’t have a strong opinion here, but I think we should discuss pros/cons)

Mobile clients will only access endpoints under `/mobile`. Optimizations could find their way back to the core API but any changes there would not need to take the mobile app(s) into account.

As we have more discussions about the mobile app it’s becoming clear that it’s key difference from va.gov will be that it’s a distilled experience. e.g. You’re not going to fill out a multi page 526 form on the mobile app, but you will get updates as it moves through the system. This difference in how the data is consumed is why mobile apps would always use `/mobile`.

**Pros**:

- Web and mobile endpoints can serialize customized data for their views
- There will be mobile only endpoints, all the mobile endpoints will live in one place
- Easier to document the mobile API as separate entity
- Less version creep, each client manages breaking changes independently
- More consistent in the core API (versions naturally evolve, a v1 mobile endpoint doesn’t live between web’s v0 and v2)
- Easier to move the mobile API out of vets-api completely should the need arise

**Cons**:

- Although routes and controllers will be unique, services will be shared and must not break wither contract.
- More tests! Vets-api takes a while to run its test suite, will be more than if they shared endpoints.
- There’s now three entry points to the API core, lighthouse, and mobile

> How might a discovery API facilitate flexibility here?

Because mobile apps are less frequently updated (if at all for some users) it will definitely use a discovery endpoint which accepts an app version and returns a directory of routes. Similar to how [Github’s](https://api.github.com/) does. It could also help in a shared scenario mapping routes for both web and mobile. Assuming web wants a discovery API and the added complexity of looking up its routes.

> Will we have a deprecation policy for mobile endpoints, slash, how out-of-date can a user’s VA.gov mobile app be before we stop supporting it? - follow-up: assuming we do eventually stop supporting old versions of mobile apps, will that include functionality in the mobile app to alert the user that they need to upgrade? (i.e. can we control that from an API endpoint of ours?

We think deprecation requires more analysis. For discussion’s sake, let’s say we release 1.0.0 in Nov, 1.1.0 in Jan, and 1.2.0 in April. There will most likely be several point releases between each of these. We don’t imagine setting a hard line when an app version becomes unsupported, but as time goes on I’m sure we’ll notice trends. For example, if we are on 2.1.0, and there is a burden to retaining the 1.0.0 functionality in one or more of our /mobile endpoints, then we can choose a sunset date for 1.0.0 depending on the user base and activity of that version’s cohorts. Although React Native (using Expo) allows for over-the-air code updates, we have no plans to use that feature. However, the discovery endpoint can be used to inform an app user that their experience will be diminished and they should attempt to upgrade to 1.1.0.

Reply:
Thanks for the answers, that all sounds great. Can you work those into the design doc for posterity?

re: `/mobile` endpoints, SGTM 👍 

re: discovery API, 👍 

re: deprecation policy:

> the discovery endpoint can be used to inform an app user that their experience will be diminished and they should attempt to upgrade to 1.1.0.

cool, I think this is the part I was hoping for in particular